### PR TITLE
:memo: Add Client assertions page

### PR DESCRIPTION
**Pull request by quilicicf**

### CODE REVIEW

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+Restlet Client allows you to test the behavior of your API.
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md]: "APIs" plural sounds better here IMO

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+To do so, you can validate the expected characteristics of an HTTP response to your requests via _Assertions_.
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md]: I find the phrase a bit complex and there is a grammar issue as "response" is singular and "requests" plural.

There is also redundance with the following sentence IMO. 

We could fuse them all into one little paragraph maybe?

"Restlet Client allows you to test the behavior of your APIs.
This is achieved by using __Assertions__ to verify that a certain part of an HTTP response matches your expectations. You can define assertions against the following response characteristics:"

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+Assertions allow you to verify that a certain part of the response matches your expectations. You can choose between:
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md]: If following my suggestion, can remove this line.

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+* Status code
+* Headers
+* Body length
+* Duration
+* Status message
+* JSON body: use this if the response returns a JSON body. You can then use [JSON path](./json-path) to verify a
+specific sub-part of the response body.
+* XML body: use this if the response returns an XML body. You can then use [Xpath](./xpath) to verify a specific
+sub-part of the response body.
+* Body content: use this if you want to assert something on the body, it will be considered as a simple string.
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md]: Not super clear here what will be "considered a simple string". 

Maybe do a full stop then   ". The body will be considered as a simple string"

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+Assertions are validated in real-time, meaning that they are re-evaluated when you run the request and when you
+update them.
+
+<!-- IN SCREENSHOT: RP_ASSERTIONS -->
+![assertions](./images/assertions.png)
+
+An assertion leverages an operator in order to compare a value to the expected one.
+Operators allow you to compare values, check existence and check inclusions.
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md]: "Compare" is repeated twice. Maybe could fuse into one sentence? 

I.e. "Assertions leverage operators in order to compare values, check their existence or inclusion of certain values".

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+Here are some examples:
+
+* "Status code" equals 200
+* "Duration (ms)" less than 200
+* "Body content" exists
+* "JSON body" identified by $.id equals 12345
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md]: Je ne comprends pas trop le "identified by", on dirait que c'est le nom d'un opérateur mais je ne vois rien qui correspond dans Client. Peut-être à clarifier?

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+Go have a look at the [__full operators list__](./assertions).
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md]: "Go have a look" c'est un peu brutal, genre "va regarder là bas" :D 

Peut-être plus doux serait un truc du genre: "Feel free to have a look at ..."

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+<a class="anchor" name="unleash-the-full-power-of-api-testing"></a>
+## Unleash the full power of API testing
+
+Assertions are good and are a solid base to test your API but you'll soon find out that you need them to be __dynamic__.
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md]: "Assertions are good" sounds a bit weird IMO. What does that mean? Maybe jump straight to "Assertions are a solid base for testing your APIs but..."

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
+Restlet Client allows you to validate the behavior of your API by creating assertions on a request.
+Assertions are based on a certain part of the response and check that the said part passes some tests.
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md]: The "said part" phrasing is probably IMO a bit too literary and complex for a general user guide. 

Perhaps we can simplify with something like:

"An assertion will verify a specific part of a response based on the conditions you've defined and will indicate whether or not the conditions pass."

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+  * [Length](#length)
+  * [Content](#content)
+  * [Regular expression matching](#regular-expression-matching)
+
+<!-- TOC: END -->
+
+<a class="anchor" name="usage-example"></a>
+## Usage example
+
+I want to make sure my content-negotiation is working. I need to know whether the header `Content-type` of my HTTP
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md]: "whether the 'Content-type' header of my ..."

Comment (jmcx) [reply modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md]: "is equal to the 'Accept' header of the ..."

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+response is equal to the header `Accept` of the request.
+
+I can do that by creating the following assertion:
+`Header > Content-type > Equals > ${"My request"."request"."headers"."Accept"}`
+
+> Note: the last part is an __expression__, it allows you to grab a part of the request to check it against the response
+in that case but it can do __much more__. Have a look at the
+[documentation](../make-your-requests-and-assertions-dynamic/expressions).
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md]: Phrasing of the note gets a tad awkward IMO starting from "in that case".

Could simplify slightly by saying:

"... to check it against the response. Please view the full documentation on expressions for more details as to what can be achieved with them."

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+<a class="anchor" name="assertion-operators"></a>
+## Assertion operators
+
+Assertions rely on operators - the part `Equals` in the example above. Restlet Client has a set of operators that you
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md]: "the 'Equals' part in the ..."

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+can use that are described below.
+
+> Note: JSON data types are only usable with the response element `JSON body`.
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md]: No super clear IMO, Do you mean "If your response body is in JSON, you'll need to use the 'JSON body' response element in order to define assertions against the values in the response body."

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+
+<a class="anchor" name="equality"></a>
+### Equality
+
+The operator `Equals` and `Does not equal` are the simplest of all and only do an equality check.
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md]: "The operator" -> "Operators"

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+* Greater than or equal
+* Less than
+* Less than or equal
+
+Comparisons are done this way:
+
+* On numbers: does a number comparison
+* On strings: compares the string lengths
+* On booleans, JSON arrays, JSON objects: compares them as strings
+* On null: N/A
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md]: Seeing "N/A" makes me wonder what will Client do? will it ignore the assertion? if so probably worth mentioning here ?

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+<a class="anchor" name="length"></a>
+### Length
+
+The length operators are:
+
+* Length equals
+* Length greater than or equal
+* Length less than or equal
+
+The length is computed this way:
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md]: "this way" -> "as follows"

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+
+* On strings: it returns the number of characters in the string
+* On JSON arrays: it returns the number of elements in the array
+* On JSON objects: it returns the number of keys in the object
+* On null, numbers, booleans: N/A
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md]: same comment about N/A

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+* On strings: it checks that the element contains the given substring.
+  * `Darth Vader` contains `arth`
+* On JSON arrays: it checks that the element contains all of the given values. The expected values can be either a
+JSON array or comma-separated values.
+  * `[ "Darth Vader", "Luke Skywalker", "Leia Skywalker" ]` contains `"Darth Vader", "Luke Skywalker"`
+  * `[ "Darth Vader", "Luke Skywalker", "Leia Skywalker" ]` contains `[ "Darth Vader" ]`
+* On JSON objects: it checks that the element contains all of the given keys.
+  * `{ "age": 41, "lightSaberColor": "red", "forcePower": "3000" }` contains `"age","lightSaberColor"`
+  * `{ "age": 41, "lightSaberColor": "red", "forcePower": "3000" }` contains `[ "forcePower" ]`
+* On null, numbers and booleans: N/A
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md]: same again

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+
+<a class="anchor" name="regular-expression-matching"></a>
+### Regular expression matching
+
+The operator `Matches` allows you to make sure an element matches a given regular expression.
```
Comment (jmcx) [inline modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md]: "make sure" -> "verify that"

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md (hunk 6 above)
Comment (quilicicf) [reply modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md]: C'est un JSON path pour trouver une partie du JSON body.

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md (hunk 13 above)
Comment (quilicicf) [reply modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md]: En fait, si tu as sélectionné le type JSON body, on traite l'input comme du JSON, sinon on le traite comme du texte.

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md (hunk 15 above)
Comment (quilicicf) [reply modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md]: It displays an error.

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md (hunk 6 above)
Comment (jmcx) [reply modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md]: Okay donc "identified by" fait partie de la syntaxe JSON path?

Comment (jmcx) [reply modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md]: Si c'est le cas, tant mieux c'est bon du coup 👍

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md (hunk 13 above)
Comment (jmcx) [reply modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/01_assertions.md]: Note: if you want a response body to be interpreted as JSON, you need to use the "JSON body" element in your assertion. If you use the "Body content" element, the body will be interpreted as a string.

File: modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md (hunk 6 above)
Comment (jmcx) [reply modules/com.restlet/tech-doc/client/guide/03_test/02_validate-http-responses/00_introduction.md]: Element "id" identified by "$.id" in the "JSON body" equals 12345

**Pull request merged**